internal/providers/vault/certificate.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package vault
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	vaultclient "github.com/hashicorp/vault-client-go"
+	"github.com/hashicorp/vault-client-go/schema"
+	"namespacelabs.dev/foundation/framework/secrets"
+	"namespacelabs.dev/foundation/framework/secrets/combined"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+const (
+	VaultJwtAudience    = "vault.namespace.systems"
+	vaultRequestTimeout = 10 * time.Second
+)
+
+func Register() {
+	p := &provider{}
+	combined.RegisterSecretsProvider(
+		func(ctx context.Context, srvRef *secrets.ServerRef, cfg *vault.Certificate) ([]byte, error) {
+			ca := cfg.GetCa()
+			if ca == nil {
+				return nil, fnerrors.BadInputError("invalid vault certificate configuration: missing CA configuration")
+			}
+
+			if err := p.Login(ctx, ca, VaultJwtAudience); err != nil {
+				return nil, err
+			}
+
+			commonName := fmt.Sprintf("%s.%s", strings.ReplaceAll(srvRef.RelPath, "/", "-"), cfg.BaseDomain)
```

Comment:
what about multi-repo?
Rel path is not enough as a unique identifier.

Reply:
tbh, there is no strong requirement of having unique server name here, but rather unique CN per region, which ensured with combination of server name + `base_domain`.
Ideally I want to to read server name (`name` in `server` spec) using server ref, but not sure if possible.